File: src/checks/usage.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

export const trackUsageCheck: Check = {
  name: 'Track Usage',
  description: 'Track usage without billing (internal visibility)',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    // Prefer testCustomerId (funded) for tests, fall back to created customer
    const customerId = ctx.testCustomerId || ctx.createdCustomerId;

    if (!customerId) {
      return {
        name: 'Track Usage',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      const result = await client.trackUsage({
        customerId,
        meter: 'api_call',
        quantity: 1,
        description: 'Health check test usage',
        units: 'requests',
      });
      const duration = performance.now() - start;

      return {
        name: 'Track Usage',
        success: result.success,
        duration,
        message: `${result.usageEventId} (${result.message})`,
        details: `Usage ID: ${result.usageEventId}, Type: ${result.usageType}, Quantity: ${result.quantity}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      return {
        name: 'Track Usage',
        success: false,
        duration,
        message: err.code || err.message || 'Failed to track usage',
        details: `${err.message} (code: ${err.code}, status: ${err.statusCode})`,
      };
    }
  },
};

export const trackUsageIdempotencyCheck: Check = {
  name: 'Track Usage Idempotency',
  description: 'Verify idempotency key deduplication in trackUsage',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.testCustomerId || ctx.createdCustomerId;

    if (!customerId) {
      return {
        name: 'Track Usage Idempotency',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    const idempotencyKey = `health-check-usage-idem-${Date.now()}-${Math.random().toString(36).slice(2)}`;

    try {
      // First call with idempotency key
      const firstResult = await client.trackUsage({
        customerId,
        meter: 'api_call',
        quantity: 5,
        description: 'Idempotency test usage',
        units: 'requests',
        idempotencyKey,
      });

      // Second call with same idempotency key - should return same event
      const secondResult = await client.trackUsage({
        customerId,
        meter: 'api_call',
        quantity: 5,
        description: 'Idempotency test usage',
        units: 'requests',
        idempotencyKey,
      });

      const duration = performance.now() - start;

      // Both calls should return the same usage event ID
      if (firstResult.usageEventId === secondResult.usageEventId) {
        return {
          name: 'Track Usage Idempotency',
          success: true,
          duration,
          message: 'Duplicate prevented (same event ID)',
          details: `Key: ${idempotencyKey}, Event ID: ${firstResult.usageEventId}`,
        };
      } else {
        return {
          name: 'Track Usage Idempotency',
          success: false,
          duration,
          message: 'Duplicate usage created',
          details: `First: ${firstResult.usageEventId}, Second: ${secondResult.usageEventId}`,
          suggestion: 'Idempotency key was not respected - duplicate usage events were created',
        };
      }
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;
      return {
        name: 'Track Usage Idempotency',
        success: false,
        duration,
        message: err.message || 'Idempotency check failed',
        details: err.code,
      };
    }
  },
};

File: python/checks/usage.py
"""Usage tracking check."""
import time
import uuid
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _track_usage_check(ctx: CheckContext) -> CheckResult:
    """Track usage without charging."""
    customer_id = ctx.created_customer_id or ctx.test_customer_id
    if not customer_id:
        return CheckResult(
            name="track_usage",
            success=False,
            duration=0,
            message="No customer ID available"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'track_usage'):
            return CheckResult(
                name="track_usage",
                success=True,
                duration=0,
                message="Usage tracking not available in SDK",
                details="Skipping usage tracking test"
            )

        result = client.track_usage(
            customer_id=customer_id,
            meter="tokens",
            quantity=500,
            units="tokens",
            description="Health check usage tracking"
        )

        return CheckResult(
            name="track_usage",
            success=True,
            duration=0,
            message="Usage tracked successfully",
            details="quantity: 500 tokens"
        )
    except Exception as e:
        return CheckResult(
            name="track_usage",
            success=False,
            duration=0,
            message=f"Usage tracking failed: {e}"
        )


track_usage_check = Check(
    name="track_usage",
    description="Track usage without charging",
    run=_track_usage_check
)


async def _track_usage_idempotency_check(ctx: CheckContext) -> CheckResult:
    """Verify idempotency key deduplication in track_usage."""
    customer_id = ctx.created_customer_id or ctx.test_customer_id
    if not customer_id:
        return CheckResult(
            name="track_usage_idempotency",
            success=False,
            duration=0,
            message="No customer ID available"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'track_usage'):
            return CheckResult(
                name="track_usage_idempotency",
                success=True,
                duration=0,
                message="Usage tracking not available in SDK",
                details="Skipping idempotency test"
            )

        idempotency_key = f"health-check-usage-idem-{int(time.time())}-{uuid.uuid4().hex[:8]}"

        # First call with idempotency key
        result1 = client.track_usage(
            customer_id=customer_id,
            meter="api_calls",
            quantity=5,
            units="requests",
            description="Idempotency test usage",
            idempotency_key=idempotency_key,
        )

        # Second call with same idempotency key - should return same event
        result2 = client.track_usage(
            customer_id=customer_id,
            meter="api_calls",
            quantity=5,
            units="requests",
            description="Idempotency test usage",
            idempotency_key=idempotency_key,
        )

        event_id_1 = getattr(result1, 'usage_event_id', str(result1))
        event_id_2 = getattr(result2, 'usage_event_id', str(result2))

        if event_id_1 == event_id_2:
            return CheckResult(
                name="track_usage_idempotency",
                success=True,
                duration=0,
                message="Duplicate prevented (same event ID)",
                details=f"Key: {idempotency_key}, Event ID: {event_id_1}"
            )
        else:
            return CheckResult(
                name="track_usage_idempotency",
                success=False,
                duration=0,
                message="Duplicate usage created",
                suggestion=f"First: {event_id_1}, Second: {event_id_2}"
            )
    except Exception as e:
        return CheckResult(
            name="track_usage_idempotency",
            success=False,
            duration=0,
            message=f"Idempotency check failed: {e}"
        )


track_usage_idempotency_check = Check(
    name="track_usage_idempotency",
    description="Verify idempotency key deduplication in track_usage",
    run=_track_usage_idempotency_check
)
